image/png.ksy

```diff
@@ -292,6 +297,56 @@ types:
       - id: text_datastream
         process: zlib
         size-eos: true
+  animation_control_chunk:
+    doc-ref: https://wiki.mozilla.org/APNG_Specification#.60acTL.60:_The_Animation_Control_Chunk
+    seq:
+      - id: num_frames
+        type: u4
+        doc: Number of frames
+      - id: num_plays
+        type: u4
+        doc: Number of times to loop, 0 indicates infinite looping.
+  frame_control_chunk:
+    doc-ref: https://wiki.mozilla.org/APNG_Specification#.60fcTL.60:_The_Frame_Control_Chunk
+    seq:
+      - id: sequence_number
+        type: u4
+        doc: Sequence number of the animation chunk
+      - id: width
+        type: u4
+        doc: Width of the following frame
+      - id: height
+        type: u4
+        doc: Height of the following frame
+      - id: x_offset
+        type: u4
+        doc: X position at which to render the following frame
+      - id: y_offset
+        type: u4
+        doc: Y position at which to render the following frame
+      - id: delay_num
+        type: u2
+        doc: Frame delay fraction numerator
+      - id: delay_den
+        type: u2
+        doc: Frame delay fraction denominator
```

Comment:
These two values actually form one value, so it would be reasonable to create a [value instance](https://doc.kaitai.io/user_guide.html#_value_instances) that will calculate the final value of `delay` from the values of `delay_num` and `delay_den`.

If this pattern of a fractional value would repeat, I'd recommend to create a self-contained subtype for it. You can do it here too, but it's not necessary - if you prefer to stick to the specification as closely, it's OK and just add the value instance for convenient access.

Reply:
I have not seen it in any other PNG chunk that I have encountered so far. I will look into the value instance

Reply:
I am at a bit at a loss of how to do this properly, as there is a special case in case delay_den is 0. I don't know how to capture that in kaitai (still learning).

Reply:
> as there is a special case in case delay_den is 0

Are you looking for the [ternary (conditional) operator](https://doc.kaitai.io/user_guide.html#_ternary_if_then_else_operator)?

If you have any struggles with putting the expression together, just describe what you need and we'll help you.

Reply:
It is this part of the specification:

```
The `delay_num` and `delay_den` parameters together specify a fraction indicating the time to display the current frame, in seconds. If the denominator is 0, it is to be treated as if it were 100 (that is, `delay_num` then specifies 1/100ths of a second). If the the value of the numerator is 0 the decoder should render the next frame as quickly as possible, though viewers may impose a reasonable lower bound. 
```

So in pseudo code it would be something like this:

if delay_den == 0, then delay_den = 100
value = delay_num/delay_den

Reply:
> if delay_den == 0, then delay_den = 100
> value = delay_num/delay_den

That's an easy one:

```ksy
instances:
  delay:
    value: delay_num / (delay_den == 0 ? 100.0 : delay_den)
```

The only thing that you need to be aware of is that given that both `delay_num` and `delay_den` are **integers**, you need to convert one of the division operands to a float (`double`) in order to get _float division_ (not integer division, which happens if you do `A / B` with both `A` and `B` integers). See https://doc.kaitai.io/user_guide.html#_arithmetic_operators:

> <table>
> <tbody>
> <tr>
> <th>NOTE</th>
> <td>If both operands are integers, the result of an arithmetic operation is an integer, otherwise it is a floating point number. For example, that means that <code>7 / 2</code> is <code>3</code>, and <code>7 / 2.0</code> is <code>3.5</code>.</td>
> </tr>
> </tbody>
> </table>

Kaitai Struct has a strong type system under the hood and it tracks type of every symbol in all expressions. Here, I used the fact that you can make the overall derived type of a ternary operation (`condition ? if_true : if_false`) a float by making either `if_true` or `if_false` a float - I deliberately use `100.0` (instead of just `100`) so that the division works as intended. If there was just `100`, the result would be truncated to an integer (you could just get `0` seconds or `1` second, but not anything between that).

Reply:
OK, I added it, including the doc as requested. I hope that everything is fine now. I tested it with the web IDE and it seems to work.